### ISSUE TITLE
Compilation failure in syscallbuf.c on incomplete struct f_owner_ex

### ISSUE THREAD

**Issue by devnull0000** [mozilla/rr#2693]
So I just tries to build rr to check how nice it will work with latest patches for AMD (I have  AMD Ryzen 9 3950X)
unfortunately I faced a compilation error which says something like:

```
make  -f CMakeFiles/breakpoint_32.dir/build.make CMakeFiles/breakpoint_32.dir/depend
/home/tot/proj/rr/src/preload/syscallbuf.c:1284:18: error: invalid application of ‘sizeof’ to incomplete type ‘struct f_owner_ex’
 1284 |     ptr += sizeof(*owner2);
```

I was able to fix it applying the following patch (just adding fcntl.h header and removing two others)

```
diff --git a/src/preload/syscallbuf.c b/src/preload/syscallbuf.c
index f8817dbf..9f9f07a4 100644
--- a/src/preload/syscallbuf.c
+++ b/src/preload/syscallbuf.c
@@ -52,8 +52,8 @@
 #include <asm/poll.h>
 #include <asm/signal.h>
 #include <asm/siginfo.h>
-#include <asm/stat.h>
 #include <asm/statfs.h>
+#include <fcntl.h>
 #include <linux/eventpoll.h>
 #include <linux/futex.h>
 #include <linux/fcntl.h>
@@ -71,7 +71,6 @@
 #include <linux/stat.h>
 #include <linux/time.h>
 #include <linux/types.h>
-#include <linux/uio.h>
 #include <linux/un.h>
 #include <stdarg.h>
 #include <stdio.h>
```

I don't know for sure but I suspect it depends on glibc version and FWIW I have it 2.32 (I'm on gentoo linux). Also I have gcc-10.2, linux-headers 5.8 and kernel 5.8.

I do not state my patch can be safely applied without breaking other systems but it helped me.

**Comment by khuey** [comment]
`f_owner_ex` is defined in kernel headers so I don't see why glibc version would matter. Is `f_owner_ex` not defined in `/usr/include/asm-generic/fcntl.h` on your system? What happens if you include `asm/fcntl.h` directly?

**Comment by devnull0000** [comment]
asm/fcntl.h expands as:
#include <asm-generic/fcntl.h>
which in turn points to:
/usr/include/asm-generic/fcntl.h
which has structure with name __kernel_f_owner_ex
however note this __kernel_ prefix

the file in turn belongs to
sys-kernel/linux-headers-5.8 (/usr/include/asm-generic/fcntl.h) 
package

after learning the stuff around I found that it's most likely gentoo specific and is a consequence of fixing the following bugs:
https://bugs.gentoo.org/388633
https://bugs.gentoo.org/244470

and the link to the patch itself 
https://gitweb.gentoo.org/proj/toolchain/linux-headers-patches.git/tree/4.16/00_all_0004-asm-generic-fcntl.h-namespace-kernel-file-structs.patch?id=b992677feb8c18f6a67629578b0c8497a5c48638

with argumentation that 
"No one should be using these structs, but just in case they are,
keep them available in the __kernel_ namespace."

Looks like the problem is known to ubuntu world as well however they there just didn't solve it:
https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1739300
and even on CentOS:
https://bugs.centos.org/view.php?id=10117

Now I suspect that if apply my patch compilation on other systems might fail (would be nice to try), and it's clear  why I feel the situation like "everything everywhere works well for everyone except me".

If you don't fix the bug in a way it would work for everybody, could you at least add some note to gentoo users, because it's a little bit confusing (and time consuming) to understand what's up.

**Comment by khuey** [comment]
Nice detective work. I did try your patch on Ubuntu and it does fail.

Maybe we should just define f_owner_ex ourselves.

**Comment by devnull0000** [comment]
Big thank you, but it's a bit not enough, because the same file (asm-generic/fcntl.h) has two more structures:

```
[ 19%] Building C object CMakeFiles/rrpreload_32.dir/32/preload/syscallbuf.c.o
/home/tot/proj/rr_build/32/preload/syscallbuf.c: In function ‘sys_fcntl64_setlk64’:
/home/tot/proj/rr_build/32/preload/syscallbuf.c:1326:18: error: invalid application of ‘sizeof’ to incomplete type ‘struct flock64’
```

to be precise flock and flock64
```
struct __kernel_flock {
        short   l_type;
        short   l_whence;
        __kernel_off_t  l_start;
        __kernel_off_t  l_len;
        __kernel_pid_t  l_pid;
        __ARCH_FLOCK_PAD
};
#endif

#ifndef HAVE_ARCH_STRUCT_FLOCK64
#ifndef __ARCH_FLOCK64_PAD
#define __ARCH_FLOCK64_PAD
#endif

struct __kernel_flock64 {
        short  l_type;
        short  l_whence;
        __kernel_loff_t l_start;
        __kernel_loff_t l_len;
        __kernel_pid_t  l_pid;
        __ARCH_FLOCK64_PAD
};
#endif
```

but only two these (no others) so making the same with them should be totally enough.

FWIW: rr now works on my AMD like a charm, and I'm just super happy.

**Comment by rocallahan** [comment]
Did you fix that locally? If you did, can you send us a PR?

**Comment by devnull0000** [comment]
Here it is, compiles and works for me:

```
diff --git a/src/preload/syscallbuf.c b/src/preload/syscallbuf.c
index b0867b53..0c43dcda 100644
--- a/src/preload/syscallbuf.c
+++ b/src/preload/syscallbuf.c
@@ -123,6 +123,18 @@ struct rr_f_owner_ex {
   int pid;
 };
 
+#ifndef __ARCH_FLOCK64_PAD
+#define __ARCH_FLOCK64_PAD
+#endif
+struct rr_flock64 {
+    short  l_type;
+    short  l_whence;
+    __kernel_loff_t l_start;
+    __kernel_loff_t l_len;
+    __kernel_pid_t  l_pid;
+    __ARCH_FLOCK64_PAD
+};
+
 /* Nonzero when syscall buffering is enabled. */
 static int buffer_enabled;
 /* Nonzero after process-global state has been initialized. */
@@ -1313,10 +1325,10 @@ static int sys_fcntl64_setlk64(const struct syscall_info* call) {
   const int syscallno = RR_FCNTL_SYSCALL;
   int fd = call->args[0];
   int cmd = call->args[1];
-  struct flock64* lock = (struct flock64*)call->args[2];
+  struct rr_flock64* lock = (struct rr_flock64*)call->args[2];
 
   void* ptr = prep_syscall_for_fd(fd);
-  struct flock64* lock2 = NULL;
+  struct rr_flock64* lock2 = NULL;
   long ret;
 
   assert(syscallno == call->no);
@@ -1347,7 +1359,7 @@ static int sys_fcntl64_setlkw64(const struct syscall_info* call) {
   const int syscallno = RR_FCNTL_SYSCALL;
   int fd = call->args[0];
   int cmd = call->args[1];
-  struct flock64* lock = (struct flock64*)call->args[2];
+  struct rr_flock64* lock = (struct rr_flock64*)call->args[2];
 
   void* ptr = prep_syscall_for_fd(fd);
   long ret;
```

**Comment by rocallahan** [comment]
Landed as f5d783dfb8eb5fe9dc5bcafdae4b559fe7342981

**Comment by devnull0000** [comment]
thanks! it works.
